src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,432 @@
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    class MultipartPipeReader
+    {
+
+        public const int DefaultHeadersCountLimit = 16;
+        public const int DefaultHeadersLengthLimit = 1024 * 16;
+        private const int DefaultBufferSize = 1024 * 4;
+        private const int StackAllocThreshold = 128;
+
+        private readonly PipeReader _pipeReader;
+        private ReadOnlySequence<byte> _buffer;
+        private ReadResult _readResult = default;
+        private readonly BufferedReadStream _stream;
+        private readonly MultipartBoundary _boundary;
+        private MultipartReaderStream _currentStream;
+
+
+        public MultipartPipeReader(string boundary, PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+            _buffer = default;
+
+            _stream = new BufferedReadStream(pipeReader.AsStream(), DefaultBufferSize);
+            _currentStream = new MultipartReaderStream(_stream, _boundary) { LengthLimit = HeadersLengthLimit };
+
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional limit for the total response body length.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+
+        public async Task<MultipartSection> ReadNextSectionAsync(CancellationToken cancellationToken = new CancellationToken())
+        {
+
+            if (_buffer.IsEmpty)
+            {
+                _readResult = await _pipeReader.ReadAsync(cancellationToken);
+                _buffer = _readResult.Buffer;
+            }
+
+            SkipMetaData(ref _buffer, _readResult.IsCompleted);
+            if(_buffer.IsEmpty)
+            {
+                return null;
+            }
+            var headersAccumulator = new KeyValueAccumulator();
+            ParseHeaders(ref _buffer, ref headersAccumulator, _readResult.IsCompleted);
+            _boundary.ExpectLeadingCrlf = true;
+            var body = ReadBody(ref _buffer, _readResult.IsCompleted);
+            return new MultipartSection() { Headers = headersAccumulator.GetResults(), Body = new MemoryStream(body), BaseStreamOffset = 0 };
+        }
+
+
+        void SkipMetaData(ref ReadOnlySequence<byte> buffer, bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            var consumedBytes = default(long);
+            var crlfDelimiter = new byte[] { (byte)'\r', (byte)'\n' };
```

Comment:
```suggestion
            ReadOnlySpan<byte> crlfDelimiter = new byte[] { (byte)'\r', (byte)'\n' };
```
Use C#-compiler's ability to refer directly to static data-

Reply:
Will do in next commit